File: src/web/magmac/app/lang/web/TypescriptRules.ts
import { Lists } from "../../../../magmac/api/collect/list/Lists";
import { LocatingRule } from "../../../../magmac/app/compile/rule/LocatingRule";
import { NodeListRule } from "../../../../magmac/app/compile/rule/NodeListRule";
import { NodeRule } from "../../../../magmac/app/compile/rule/NodeRule";
import { OrRule } from "../../../../magmac/app/compile/rule/OrRule";
import { PrefixRule } from "../../../../magmac/app/compile/rule/PrefixRule";
import { Rule } from "../../../../magmac/app/compile/rule/Rule";
import { StringRule } from "../../../../magmac/app/compile/rule/StringRule";
import { StripRule } from "../../../../magmac/app/compile/rule/StripRule";
import { SuffixRule } from "../../../../magmac/app/compile/rule/SuffixRule";
import { TypeRule } from "../../../../magmac/app/compile/rule/TypeRule";
import { DelimitedDivider } from "../../../../magmac/app/compile/rule/divide/DelimitedDivider";
import { DelimitedFolder } from "../../../../magmac/app/compile/rule/fold/DelimitedFolder";
import { CommonLang } from "../../../../magmac/app/lang/CommonLang";
import { CommonRules } from "../../../../magmac/app/lang/CommonRules";
import { JavaRules } from "../../../../magmac/app/lang/JavaRules";
import { LazyRule } from "../../../../magmac/app/lang/LazyRule";
import { MutableLazyRule } from "../../../../magmac/app/lang/MutableLazyRule";
import { OptionNodeListRule } from "../../../../magmac/app/lang/OptionNodeListRule";
import { JavaLang } from "../../../../magmac/app/lang/java/JavaLang";
import { JavaStructureStatement } from "../../../../magmac/app/lang/java/JavaStructureStatement";
import { FunctionSegments } from "../../../../magmac/app/lang/node/FunctionSegments";
import { Modifier } from "../../../../magmac/app/lang/node/Modifier";
import { Parameters } from "../../../../magmac/app/lang/node/Parameters";
export class TypescriptRules {
	createRule() : Rule {return new TypeRule( 0, 0.Statements( 0, new OrRule( 0.of( 0.createWhitespaceRule( ), 0( ), 0( 0), 0( 0)))));;}
	createStructureMemberRule() : Rule {break;break;return new OrRule( 0.of( 0.createWhitespaceRule( ), 0( 0, 0), 0.createStructureStatementRule( 0, 0)));;}
	createMethodRule(definition : Rule, valueLazy : LazyRule) : Rule {break;break;break;break;break;return new TypeRule( 0, new OptionNodeListRule( 0, 0, new SuffixRule( 0, 0)));;}
	createConstructorRule(definition : Rule) : Rule {break;return new TypeRule( 0, new PrefixRule( 0, new SuffixRule( 0, 0)));;}
	createDefinitionRule() : Rule {break;break;break;break;break;break;return 0.set( new OptionNodeListRule( 0, 0.Last( 0, 0, 0), 0));;}
	createTypeRule() : Rule {break;return 0.set( new OrRule( 0.of( 0.JavaTemplateType.createTemplateRule( 0), 0( 0), 0.createSymbolRule( ))));;}
	createStructureRule(type : String) : Rule {break;break;break;return new TypeRule( 0, new PrefixRule( 0, 0));;}
	createImportRule() : Rule {break;break;break;return new TypeRule( 0, new PrefixRule( 0, 0));;}
	createArrayRule(orRule : LazyRule) : TypeRule {return new TypeRule( 0, new SuffixRule( new NodeRule( 0, 0), 0));;}
}

File: src/java/magmac/app/lang/web/TypescriptRules.java
package magmac.app.lang.web;

import magmac.api.collect.list.Lists;
import magmac.app.compile.rule.LocatingRule;
import magmac.app.compile.rule.NodeRule;
import magmac.app.compile.rule.OrRule;
import magmac.app.compile.rule.PrefixRule;
import magmac.app.compile.rule.Rule;
import magmac.app.compile.rule.StringRule;
import magmac.app.compile.rule.StripRule;
import magmac.app.compile.rule.SuffixRule;
import magmac.app.compile.rule.TypeRule;
import magmac.app.lang.CommonLang;
import magmac.app.lang.CommonRules;
import magmac.app.lang.JavaRules;
import magmac.app.lang.LazyRule;
import magmac.app.lang.MutableLazyRule;
import magmac.app.lang.OptionNodeListRule;
import magmac.app.lang.java.JavaLang;
import magmac.app.lang.java.JavaStructureStatement;
import magmac.app.lang.node.FunctionSegments;
import magmac.app.lang.node.Modifier;
import magmac.app.lang.node.Parameters;

public class TypescriptRules {
    public static Rule createRule() {
        return new TypeRule("root", CommonLang.Statements("children", new OrRule(Lists.of(
                JavaRules.createWhitespaceRule(),
                TypescriptLang.TypeScriptImport.createImportRule(),
                TypescriptLang.TypescriptStructureNode.createStructureRule("class"),
                TypescriptLang.TypescriptStructureNode.createStructureRule("interface")
        ))));
    }

    public static Rule createStructureMemberRule() {
        Rule definitionRule = createDefinitionRule();
        LazyRule valueLazy = new MutableLazyRule();
        return new OrRule(Lists.of(
                JavaRules.createWhitespaceRule(),
                createMethodRule(definitionRule, valueLazy),
                JavaStructureStatement.createStructureStatementRule(definitionRule, valueLazy)
        ));
    }

    private static Rule createMethodRule(Rule definition, LazyRule valueLazy) {
        Rule header = new PrefixRule("\n\t", new NodeRule("header", new OrRule(Lists.of(
                definition,
                createConstructorRule(definition)
        ))));

        LazyRule functionSegmentRule = new MutableLazyRule();
        LazyRule value = JavaRules.initValueRule(functionSegmentRule, valueLazy, " => ", definition);
        Rule children = CommonLang.Statements("children", FunctionSegments.initFunctionSegmentRule(functionSegmentRule, value, definition));
        Rule childRule = new SuffixRule(LocatingRule.First(header, " {", new StripRule("", children, "after-children")), "}");
        return new TypeRule("method", new OptionNodeListRule("children",
                childRule,
                new SuffixRule(header, ";")
        ));
    }

    private static Rule createConstructorRule(Rule definition) {
        Rule parametersRule = Parameters.createParametersRule(definition);
        return new TypeRule("constructor", new PrefixRule("constructor(", new SuffixRule(parametersRule, ")")));
    }

    private static Rule createDefinitionRule() {
        LazyRule definition = new MutableLazyRule();
        Rule modifiers = Modifier.createModifiersRule();

        Rule parameters = Parameters.createParametersRule(definition);
        Rule name = new StringRule("name");
        Rule leftRule = new OptionNodeListRule("parameters",
                new SuffixRule(LocatingRule.First(name, "(", parameters), ")"),
                name
        );

        Rule first = LocatingRule.First(leftRule, " : ", new NodeRule("type", createTypeRule()));
        return definition.set(new OptionNodeListRule("modifiers", LocatingRule.Last(modifiers, " ", first), first));
    }

    private static Rule createTypeRule() {
        LazyRule type = new MutableLazyRule();
        return type.set(new OrRule(Lists.of(
                JavaLang.JavaTemplateType.createTemplateRule(type),
                TypescriptLang.createArrayRule(type),
                CommonRules.createSymbolRule()
        )));
    }
}
